File: Algorithms_Part_2/Ex32KeywordSearch.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.Queue;
import edu.princeton.cs.algs4.StdOut;

import java.util.Arrays;

public class Ex32KeywordSearch {

    private final int length;
    private String[] suffixes;

    public Ex32KeywordSearch(String s) {
        length = s.length();
        suffixes = new String[length];
        for (int i = 0; i < length; i++) {
            suffixes[i] = s.substring(i);
            // StdOut.println(i + ":" + s.substring(i));
        }

        Arrays.sort(suffixes);
        // for (int i = 0; i < suffixes.length; i++)
        //    StdOut.println(i + ":" + suffixes[i]);
    }

    public int[] find(String query) {
        int index = getindex(query, 0, suffixes.length - 1);
        // StdOut.println("index of " + "\"" + query + "\" is " + index);
        if (index == -1) return null;
        Queue<String> ind = new Queue<>();
        while (index < suffixes.length &&
                query.equals(suffixes[index].substring(0, query.length())))
            ind.enqueue(suffixes[index++]);
        int[] indices = new int[ind.size()];
        int i = 0;
        for (String s : ind)
            indices[i++] = length - s.length();
        Arrays.sort(indices);
        return indices;
    }

    private int getindex(String query, int lo, int hi) {
        if (hi <= lo) return -1;
        int mid = lo + (hi - lo) / 2;
        String checkString = suffixes[mid].substring(0, query.length());
        // StdOut.println("query:" + query + " lo:" + lo + " hi:" + hi + " mid:" + mid + " check:"
        //                       + checkString);
        int cmp = query.compareTo(checkString);
        if (cmp < 0) return getindex(query, lo, mid);
        else if (cmp > 0) return getindex(query, mid, hi);
        else {
            while (mid > 0 && query.equals(suffixes[mid - 1].substring(0, query.length()))) mid--;
            return mid;
        }
    }

    private static void search(Ex32KeywordSearch ks, String test, String query) {
        StdOut.println("------------------------------------------------");
        StdOut.println("Search for " + "\"" + query + "\"");
        int[] indices = ks.find(query);
        int occurrences = indices != null ? indices.length : 0;
        StringBuilder builder = new StringBuilder();
        builder.append(test);
        if (occurrences > 0)
            for (int i = indices.length - 1; i >= 0; i--)
                builder.insert(indices[i], '*');
        if (occurrences == 0) StdOut.println("\"" + query + "\"" + " was not found");
        else {
            StdOut.print(
                    "\"" + query + "\"" + " was found " + indices.length + " times in indices:");
            for (int i = 0; i < indices.length; i++)
                StdOut.print(" " + indices[i]);
        }
        StdOut.println();
        StdOut.println(builder.toString());
    }

    public static void main(String[] args) {
        In in = new In("tale.txt");
        String test = in.readAll();
        Ex32KeywordSearch ks = new Ex32KeywordSearch(test);
        search(ks, test, "was");
        search(ks, test, "age");
        search(ks, test, "time");
        search(ks, test, "dark");
        search(ks, test, "ness");
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex32KeywordSearch.py
import math


class KeyWordSearch:

    def __init__(self, string):
        self.suffixes = []
        for i in range(len(string)):
            self.suffixes.append(string[i:])
        self.suffixes.sort()

    def find(self, pattern):

        def get_index(s, lo, hi):
            if hi <= lo:
                return -1
            mid = lo + int(math.floor((hi - lo) / 2))
            check = self.suffixes[mid][:len(pattern)]

            if check < s:
                return get_index(s, mid + 1, hi)
            elif s < check:
                return get_index(s, lo, mid)
            else:
                while mid > 0 and \
                        self.suffixes[mid - 1][:len(pattern)] == pattern:
                    mid -= 1
                return mid

        indices = []
        index = get_index(pattern, 0, len(self.suffixes) - 1)
        if index != -1:
            while index < len(self.suffixes) and \
                    self.suffixes[index][:len(pattern)] == pattern:
                indices.append(len(self.suffixes) - len(self.suffixes[index]))
                index += 1

        indices.sort()
        return indices


def test_key_word_search():

    with open('tale.txt') as f:
        text = f.read()

    kws = KeyWordSearch(text)

    def search(pattern):
        print()
        print('-'*50)
        print(f'Search for "{pattern}"')

        indices = kws.find(pattern)

        if len(indices) == 0:
            print(f'"{pattern}" was not found')
        else:
            print(f'"{pattern}" was found {len(indices)} times in indices: {indices}')
            builder = text
            for i in range(len(indices) - 1, -1, -1):
                builder = builder[:indices[i]] + '*' + builder[indices[i]:]
            print(builder)

    search('was')
    search('age')
    search('time')
    search('dark')
    search('ness')


if __name__ == '__main__':
    test_key_word_search()
